src/change_tracker.rs

```diff
@@ -0,0 +1,209 @@
+use core::mem;
+
+use alloc::vec::Vec;
+
+use crate::{Component, Entity, PreparedQuery, With, Without, World};
+
+/// Helper to track changes in `T` components
+pub struct ChangeTracker<T: Component> {
+    added: PreparedQuery<Without<&'static T, &'static Previous<T>>>,
+    changed: PreparedQuery<(&'static T, &'static mut Previous<T>)>,
+    removed: PreparedQuery<Without<With<(), &'static Previous<T>>, &'static T>>,
+
+    added_components: Vec<(Entity, T)>,
+    removed_components: Vec<Entity>,
+}
+
+impl<T: Component> ChangeTracker<T> {
+    /// Create a change tracker for `T` components
+    pub fn new() -> Self {
+        Self {
+            added: PreparedQuery::new(),
+            changed: PreparedQuery::new(),
+            removed: PreparedQuery::new(),
+
+            added_components: Vec::new(),
+            removed_components: Vec::new(),
+        }
+    }
+
+    /// Determine the changes in `T` components in `world` since the previous call
+    ///
+    /// For each entity with a `T` component, a private component is inserted which stores the value
+    /// as of the most recent call to `track`. This provides robust, exact change detection at the
+    /// cost of visiting each possibly-changed entity. It is a good fit for entities that will
+    /// typically be visited regardless, and `T`s having fast [`Clone`] and [`PartialEq`] impls.
+    ///
+    /// Always use exactly one `ChangeTracker` per [`World`] per component type of interest. Using
+    /// multiple trackers of the same `T` on the same world, or using the same tracker across
+    /// multiple worlds, will produce unpredictable results.
+    pub fn track<'a>(&'a mut self, world: &'a mut World) -> Changes<'a, T>
```

Comment:
Is the exclusive reference to the `World` required for correctness? Wouldn't a shared reference suffice but allow interleaving the tracking of changes to multiple types `T` on a given `World`?

Reply:
The exclusive reference is required to insert/remove `Previous<T>`s as needed for the added/removed cases. Coupling those updates tightly with the queries makes it very difficult to inadvertently double-count a change.